0005-ordered-quantification.md

```diff
@@ -0,0 +1,661 @@
+
+# Prior art
+[prior-art]: #prior-art
+
+Sequence comprehensions bear a strong resemblance to list comprehensions in Python, which also include similar abilities to 
+bind multiple variables and filter the enumerated values: `[(x, y) for x in [1,2,3] for y in [3,1,4] if x != y]`.
+Haskell and Cryptol also feature similar constructs.
+
+Although the syntax `for[each] x in c` is more common, there is precedent for using the symbol `<-` instead in other languages,
+including Scala, Cryptol, and Haskell. Using something other than `in` is recommended to avoid confusion with the `in` boolean operator,
+but I don't have a strong opinion about which symbol or keyword we use.
+
+Most mainstream programming languages include rich libraries for working with collections and iterators,
+and define much the same operations such as "filter" and "flatMap".
+The proposed Dafny features provide similar functionality at a higher level
+of abstraction that will require less effort from Dafny programmers to successfully verify: 
+quantification expressions are used to succinctly declare the behavior of
+an operation that in spirit applies several such operations. See the "Future Possibilities" section for ideas for
+pushing this philosophy even further.
+
+The [JMatch language](https://www.cs.cornell.edu/andru/papers/padl03.pdf), an extension of Java to support
+pattern matching, includes a `foreach` loop remarkably similar to the proposal here: it is parameterized by
+a boolean expression and iterates through all matching bindings. For example, the statement `foreach(a[int i] != 0) n += i;`
+sums the indices of all non-zero elements of the array `a`.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Tracking the values enumerated so far and/or the number of iterations in a `foreach` loop
+is possible with manual helper functions as illustrated above, but only when the source collection is a sequence.
+It may be a good idea in the next iteration of the feature to have convenient mechanisms for this,
+if they are frequently requested. They may be best addressed when adding support for 
+user-defined collection types and enumerators instead. Another possibility is supporting
+parallel orderings as in Cryptol or Haskell, such that indexing an arbitrary collections could be expressed as
+`seq x <- s (some symbol) i <- (0 to *) :: (s, i)`, but I don't see an obvious choice of syntax.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+Adding this building block to the Dafny language opens up a wide array of tempting new features!
+
+## Pattern matching on quantifier variables
+
+The LHS of `<-` declarations could be extended to allow case patterns, just as variable declaration statements do.
+This would allow destructuring datatypes and tuples, as in `(k, v) <- myMap.Items`.
+It could also support filtering, as in `Some(x) <- mySetOfOptionals`.
```

Comment:
I like this and I think we should do it (after the more basic functionality has been implemented and loved). Just a note: it will complicate the lookahead disambiguation that will need to be written for Coco (see my comments regarding the `","` above).

Reply:
I very much like this, too.